### DISCUSSION TITLE
Add `X-Remix-Referer` header

### DISCUSSION THREAD

**Discussion by ryanflorence** [remix-run/remix#5457] [Proposals]
There are times apps want to reference `request.headers.get("Referer")` but it's not always accurate because of the design of `popstate` events (until we have the Navigation API).

You can rely on the referer header for:

- document requests
- push navigation

But you can't rely on it for pop navigation. That's because the browser uses the current URL in the address bar for the referrer when a `fetch` is made by Remix. Remix doesn't know to the make the fetch until the pop has already changed the URL to the destination, so the destination becomes the fetch requests "referer".

We can add `X-Remix-Referer` to our fetch request for apps to use.

```jsx
export function action({ request }) {
  let referer =
    // first check for the remix header for client navigations
    request.headers.get("X-Remix-Referer") ||
    // then for document requests
    request.headers.get("Referer");
  await whatever();
  return redirect(new URL(referer).pathname);
}
```

**Comment by kiliman** [comment]
Also, some browsers will let users hide the `Referer` header for privacy reasons.

+1 for Remix adding more headers

**Comment by machour** [comment]
We could also provide a `redirectBack(defaultPath: string, init?: ResponseInit)` helper around this header. 

Use the referer value, or defaultPath if the header does not exist. 

Could also be implemented in user land or remix-utils, but worth a thought.

**Reply by ryanflorence** [reply]
absolutely, this is the use case where it pops up most often for me

**Reply by sergiodxa** [reply]
The `redirectBack` helper already exists in Remix Utils package, it uses the Referer header but if not found use the fallback path, the usage is different tho
```ts
declare function redirectBack(request: Request, init: ResponseInit & { fallback: string }): Response
```
This is because the function needs to access the Request in order to get the Referer header, and the fallback is (the default path) is passed as an extension of ResponseInit.

If `x-Remix-Referer` is implemented I could update `redirectBack` to check first for that header and only if it's not there check to `Referer` and then go to the fallback.

**Comment by bondepike** [comment]
This would be good for my application. I have a few global pages that are linked to from multiple places. But redirecting back to their source is not easy. Sometimes I navigate them back through some abort / go-back Link tag (that they click), other times the navigation is after some action.

Right now I capture their source location like this code sample.

/routes/.../some-deep-nested-route.js
```jsx
const location = useLocation()

return (
    <div>
        <Link to={`/some-global-page?src=${encodeURIComponent(location.pathname)}`}>
            I want to inspect something, then come back here
        </Link>
        ...
    </div>
)
```
